unit_test/tstNeighbor.hpp

```diff
@@ -0,0 +1,375 @@
+                Cabana::NeighborList<ListType>::numNeighbor( list, p );
+            for ( int n = 0; n < list_host.counts( p ); ++n )
+                list_host.neighbors( p, n ) =
+                    Cabana::NeighborList<ListType>::getNeighbor( list, p, n );
+        } );
+    Kokkos::fence();
+    return createTestListHostCopy( list_host );
+}
+
+//-------------------------------------------------------------------------//
+// Build a neighbor list with a brute force n^2 implementation.
+template <class PositionSlice>
+TestNeighborList<TEST_MEMSPACE>
+computeFullNeighborList( const PositionSlice &position, const double cutoff )
+{
+    // Count first.
+    TestNeighborList<TEST_MEMSPACE> list;
+    int total_atoms = position.size();
+    double rsqr = cutoff * cutoff;
+    list.counts = Kokkos::View<int *, TEST_MEMSPACE>( "test_neighbor_count",
+                                                      total_atoms );
+    Kokkos::deep_copy( list.counts, 0 );
+    auto count_op = KOKKOS_LAMBDA( const int i )
+    {
+        for ( int j = 0; j < total_atoms; ++j )
+        {
+            if ( i != j )
+            {
+                double dsqr = 0.0;
+                for ( int d = 0; d < 3; ++d )
+                    dsqr += ( position( i, d ) - position( j, d ) ) *
+                            ( position( i, d ) - position( j, d ) );
+                if ( dsqr <= rsqr )
+                    list.counts( i ) += 1;
+            }
+        }
+    };
+    Kokkos::RangePolicy<TEST_EXECSPACE> exec_policy( 0, total_atoms );
+    Kokkos::parallel_for( exec_policy, count_op );
+    Kokkos::fence();
+
+    // Allocate.
+    auto max_op = KOKKOS_LAMBDA( const int i, int &max_val )
+    {
+        if ( max_val < list.counts( i ) )
+            max_val = list.counts( i );
+    };
+    int max_n;
+    Kokkos::parallel_reduce( exec_policy, max_op, Kokkos::Max<int>( max_n ) );
+    Kokkos::fence();
+    list.neighbors = Kokkos::View<int **, TEST_MEMSPACE>( "test_neighbors",
+                                                          total_atoms, max_n );
+
+    // Fill.
+    auto fill_op = KOKKOS_LAMBDA( const int i )
+    {
+        int n_count = 0;
+        for ( int j = 0; j < total_atoms; ++j )
+        {
+            if ( i != j )
+            {
+                double dsqr = 0.0;
+                for ( int d = 0; d < 3; ++d )
+                    dsqr += ( position( i, d ) - position( j, d ) ) *
+                            ( position( i, d ) - position( j, d ) );
+                if ( dsqr <= rsqr )
+                {
+                    list.neighbors( i, n_count ) = j;
+                    ++n_count;
+                }
+            }
+        }
+    };
+    Kokkos::parallel_for( exec_policy, fill_op );
+    Kokkos::fence();
+    return list;
+}
+
+//---------------------------------------------------------------------------//
+template <class ListType, class PositionSlice>
+void checkFullNeighborListPartialRange( const ListType &list,
```

Comment:
I'm wondering if we can combine the full and partial test to reduce code here but I think this is OK for now as it is essentially what I implemented in Cabana

Reply:
Yeah, I should do that both here and in Cabana